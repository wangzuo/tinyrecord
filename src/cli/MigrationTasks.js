import fs from 'fs';
import path from 'path';
import moment from 'moment';

export default class MigrationTasks {
  static create(name, attributes) {
    const options = this.parseName(name);
    const attrs = this.parseAttributes(attributes);
    const ts = moment().format('YYYYMMDDHHmmss');
    const filename = `${ts}-${name}.js`;

    const module = this[options.template](name, attrs, options);
    const filepath = path.join(process.cwd(), './db/migrate', filename);

    fs.writeFileSync(filepath, module);
  }

  static createTable(name, attributes, { tableName }) {
    const module = `const { Migration } = require('tinyrecord');

class ${name} extends Migration {
  async change() {
    await this.createTable('${tableName}', {}, t => {
${attributes.map(({ name, type }) => `      t.${type}('${name}');`).join('\n')}
      t.timestamps();
    });
  }
}

module.exports = ${name};
`;

    return module;
  }

  static migration(name, attributes, { action, tableName }) {
    return ``;
  }

  static parseName(name) {
    const m1 = name.match(/^create(.+)/i);
    if (m1) {
      return { template: 'createTable', tableName: m1[1].toLowerCase() };
    }

    const m2 = name.match(/^(add|remove).*(?:to|from)(.*)/i);
    if (m2) {
      return {
        template: 'migration',
        action: m2[1].toLowerCase(),
        tableName: m2[2].toLowerCase()
      };
    }
  }

  static parseAttributes(attributes) {
    return attributes.map(attr => {
      const s = attr.split(':');
      return {
        name: s[0],
        type: s[1]
      };
    });
  }
}
