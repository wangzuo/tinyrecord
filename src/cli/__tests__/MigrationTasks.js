import MigrationTasks from '../MigrationTasks';

test('parseName', () => {
  expect(MigrationTasks.parseName('CreatePosts')).toEqual({
    template: 'createTable',
    tableName: 'posts'
  });

  expect(MigrationTasks.parseName('CreateSearchHistories')).toEqual({
    template: 'createTable',
    tableName: 'search_histories'
  });

  expect(MigrationTasks.parseName('AddTitleToPosts')).toEqual({
    template: 'migration',
    tableName: 'posts',
    action: 'add'
  });

  expect(MigrationTasks.parseName('RemoveTitleFromPosts')).toEqual({
    template: 'migration',
    tableName: 'posts',
    action: 'remove'
  });
});

test('createTable', () => {
  expect(
    MigrationTasks.createTable(
      'CreatePosts',
      [{ name: 'title', type: 'string' }],
      { tableName: 'posts' }
    )
  ).toMatchSnapshot();
});

test('migration', () => {
  expect(
    MigrationTasks.migration(
      'AddTitleToPosts',
      [{ name: 'title', type: 'string' }],
      { action: 'add', tableName: 'posts' }
    )
  ).toMatchSnapshot();

  expect(
    MigrationTasks.migration(
      'RemoveTitleFromPosts',
      [{ name: 'title', type: 'string' }],
      {
        action: 'remove',
        tableName: 'posts'
      }
    )
  ).toMatchSnapshot();
});
