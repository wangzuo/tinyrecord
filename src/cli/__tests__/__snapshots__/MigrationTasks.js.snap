// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTable 1`] = `
"const { Migration } = require('tinyrecord');

class CreatePosts extends Migration {
  async change() {
    await this.createTable('posts', {}, t => {
      t.string('title');
      t.timestamps();
    });
  }
}

module.exports = CreatePosts;
"
`;

exports[`migration 1`] = `
"const { Migration } = require('tinyrecord');

class AddTitleToPosts extends Migration {
  async change() {
    await this.addColumn('posts', 'title', 'string');
  }
}

module.exports = AddTitleToPosts;"
`;

exports[`migration 2`] = `
"const { Migration } = require('tinyrecord');

class RemoveTitleFromPosts extends Migration {
  async change() {
    await this.removeColumn('posts', 'title', 'string')
  }
}

module.exports = RemoveTitleFromPosts;"
`;
